### PR TITLE
Add some new interfaces for propagator; reorganize code

### PR DESCRIPTION
**Pull request by rhelmot**
sync https://github.com/angr/ailment/pull/133

This adds an interface to be able to retrieve the states from the beginning of each block as a propagation. It also adds a function to get all the non-TOP registers out of a state, i.e. the "results" of the propagation. It also fixes some type annotations in forward analysis. Generally, it's all good things, but in order to fix some awful circular import problems a lot of code had to be moved around. This ended badly.

### CODE REVIEW

File: angr/knowledge_plugins/propagations/states.py [hunk 1]
```diff
@@ -0,0 +1,667 @@
...
+        )
+
+        self._registers.set_state(self)
+        self._stack_variables.set_state(self)
+
+    def __repr__(self):
+        return "<PropagatorVEXState>"
+
+    @classmethod
+    def initial_state(cls, project, only_consts=False, gp=None, do_binops=True, store_tops=False, func_addr=None):
```
Comment (rhelmot) [inline angr/knowledge_plugins/propagations/states.py:248]: This is new

File: angr/knowledge_plugins/propagations/states.py [hunk 2]
```diff
@@ -0,0 +1,667 @@
...
+        # use this information to determine if stack reads after this store can be safely resolved to definitions prior
+        # to the stack read.
+        self.last_stack_store: Optional[Tuple[int, int, ailment.Stmt.Store]] = None
+        self.global_stores: List[Tuple[int, int, Any, ailment.Stmt.Store]] = []
+
+    def __repr__(self):
+        return "<PropagatorAILState>"
+
+    @classmethod
+    def initial_state(
```
Comment (rhelmot) [inline angr/knowledge_plugins/propagations/states.py:442]: This is new

**Comment by ltfish** [comment]
I rebased this branch on top of the latest master. Let's see if it passes CI!

**Comment by rhelmot** [comment]
@ltfish can you debug https://github.com/angr/angr/runs/12876386246

**Comment by ltfish** [comment]
Yes!

**Pull request merged**